### PR TITLE
bazecor: add wayland support

### PR DESCRIPTION
**Pull request by gcleroux**
## Description of changes

* Add wayland support
* Refactor `default.nix` to fit the [nixpkgs docs](https://nixos.org/manual/nixpkgs/unstable/#sec-pkgs-appimageTools)

Let met preface this PR by saying: Thank you @amesgen for your work on the Bazecor package!

This PR was meant to simply add wayland support to the current package, but looking at the source files
I tought I would refactor `default.nix` to fit the current `nixpkgs` docs linked above. If you feel like this is out of
scope for this PR, I can happily split it.

~When testing wayland, I found that the AppImage would often crash on startup.
Looking deeper into it, it looks like it is a [common issue](https://github.com/Dygmalab/Bazecor/issues/683). Since the current version of `Bazecor` is flaky *(pun very much intended)*, I chose to make my work disabled by default.~

EDIT: `Bazecor` seems fine on wayland now. Enabling by default.

Lastly, I saw on Discord that you were looking for a maintainer with a defy keyboard. I will happily throw my name in the hat if you will have me! 

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by amesgen — CHANGES_REQUESTED** [review]
Thanks!

> Lastly, I saw on Discord that you were looking for a maintainer with a defy keyboard. I will happily throw my name in the hat if you will have me!

That would be great, even more so as I don't use Wayland ATM. Feel free to add yourself as a maintainer in another commit in this or another PR!

File: pkgs/applications/misc/bazecor/default.nix [hunk 1]
```diff
@@ -1,52 +1,48 @@
...
       substituteInPlace \
-        $out/usr/lib/bazecor/resources/app/.webpack/main/index.js \
+        $out/usr/lib/${pname}/resources/app/.webpack/main/index.js \
         --replace \
           'checkUdev=()=>{try{if(c.default.existsSync(f))return c.default.readFileSync(f,"utf-8").trim()===l.trim()}catch(e){console.error(e)}return!1}' \
           'checkUdev=()=>{return 1}'
     '';
   };
+in appimageTools.wrapType2 {
+  inherit pname version src;
```
[the remark this answers is not in the document]
Comment (amesgen) [reply pkgs/applications/misc/bazecor/default.nix:23]: BTW: [`nix-diff`](https://github.com/Gabriella439/nix-diff) is a great tool to find changes like this.

File: pkgs/applications/misc/bazecor/default.nix [hunk 2]
```diff
@@ -17,34 +13,40 @@ appimageTools.wrapAppImage rec {
...
 
   # Also expose the udev rules here, so it can be used as:
   #   services.udev.packages = [ pkgs.bazecor ];
   # to allow non-root modifications to the keyboards.
 
   extraInstallCommands = ''
-    install -m 444 -D ${src}/Bazecor.desktop -t $out/share/applications
-    substituteInPlace $out/share/applications/Bazecor.desktop \
-      --replace 'Exec=Bazecor' 'Exec=bazecor'
+    ln -s $out/bin/${pname}-${version} $out/bin/${pname}
```
Comment (gvolpe) [inline pkgs/applications/misc/bazecor/default.nix:36]: Is this symlink here necessary? It fails to build for me, but it works if I remove it

File: pkgs/applications/misc/bazecor/default.nix [hunk 3]
```diff
@@ -17,34 +13,40 @@ appimageTools.wrapAppImage rec {
...
+
+    source "${makeWrapper}/nix-support/setup-hook"
+    wrapProgram $out/bin/${pname} \
```
Comment (gvolpe) [inline pkgs/applications/misc/bazecor/default.nix:39]: I would place this behind the `enableWayland` flag as initially suggested, not sure if this would affect users on Xorg

**Review by gvolpe — COMMENTED** [review]
Thanks for submitting this PR! Tested it on Hyprland, works with some caveats (removing the symlink).

File: pkgs/applications/misc/bazecor/default.nix (hunk 2 above)
Comment (gcleroux) [reply pkgs/applications/misc/bazecor/default.nix:36]: Interesting, I didn't get any errors last time I built it locally.
I will change `ln` to `mv` and won't run into this issue anymore

File: pkgs/applications/misc/bazecor/default.nix (hunk 3 above)
Comment (gcleroux) [reply pkgs/applications/misc/bazecor/default.nix:39]: > I would place this behind the enableWayland flag as initially suggested

I thought about it a while ago and it really didn't sit right with me in the end to add a new variable to the build that we would have to maintain. I quickly looked around in the `nixpkgs` repo for toher uses of the `enableWayland` input and it was never used the way I intended to use it so it felt very icky.

I updated to `zig v0.3.0` recently and bazecor with the wayland flags never crashes on startup. Since Bazecor is still on `1.3.11` and has been for a while, it looks like the crashes were a compositor issue and not related to bazecor anyway. I think the best way going forward is to enable wayland support by default and users who experience problems on their side can overlay the `wrapProgram` command instead. No need for `enableWayland` anymore.


> not sure if this would affect users on Xorg

This PR has no effect on Xorg users since they won't have the `NIXOS_OZONE_WL` var set up in their `configuration.nix`. If `NIXOS_OZONE_WL` is not defined, it wraps the `bazecor` command with nothing.

**Comment by gcleroux** [comment]
Sorry I haven't been able to contribute much time and energy to this PR since I was trying to balance work and my last uni semester these past few weeks. I updated this PR a few days ago wanting to bring wayland support for everybody but kept pushing back writing an update here on my to-do list.

Thank you @gvolpe for the review and the sudden urge to take something off the top of my list :)

File: pkgs/applications/misc/bazecor/default.nix (hunk 2 above)
Comment (gcleroux) [reply pkgs/applications/misc/bazecor/default.nix]: @gvolpe I think this PR has been gathering dust for too long and resulted in this breakage. 

`appimageTools` no longer creates a binary `$out/bin/${name}` if `pname` and `version` are provided. See #271071.

It used to be needed at the time I wrote this PR. I updated the PR to reflect these changes.

**Comment by gcleroux** [comment]
@amesgen 

TL;DR

I have been running `Bazecor` on [river](https://codeberg.org/river/river) `v0.3.0` for a few weeks now and haven't encountered a single startup crash. I'm pretty sure it was a compositor problem and not actually `Bazecor` that was responsible for the crash since it hasn't been updated in months. 

I feel like enabling the `wayland` flags by default is the way to go going forward.

**Review by gvolpe — APPROVED** [review]
Thanks!

**Review by amesgen — APPROVED** [review]
Thanks, LGTM!

(Minor: You could remove the merge commit by instead rebasing the two commits of this PR on a recent `master` commit.)

File: pkgs/applications/misc/bazecor/default.nix [hunk 4]
```diff
@@ -53,7 +53,7 @@ appimageTools.wrapAppImage rec {
     changelog = "https://github.com/Dygmalab/Bazecor/releases/tag/v${version}";
     sourceProvenance = [ lib.sourceTypes.binaryNativeCode ];
     license = lib.licenses.gpl3Only;
-    maintainers = with lib.maintainers; [ amesgen ];
+    maintainers = with lib.maintainers; [ amesgen gcleroux ];
```
Comment (amesgen) [inline pkgs/applications/misc/bazecor/default.nix:56]: Note that you need to [add youself to the `maintainer-list.nix`](https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md#how-to-become-a-maintainer); currently, ofborg is failing because you are not in there.

Comment (gcleroux) [reply pkgs/applications/misc/bazecor/default.nix]: Ah thanks, I forgot to do that part

**Comment by gcleroux** [comment]
> Thanks, LGTM!
> 
> (Minor: You could remove the merge commit by instead rebasing the two commits of this PR on a recent `master` commit.)

Yeah good idea, should've done that from the get go to be honest. It should be good now.

**Review by amesgen — APPROVED** [review]
:+1:

Minor (sorry that I didn't link to this before):

> When adding yourself as maintainer in the same pull request, make a separate commit with the message `maintainers: add <handle>`. Add the commit before those making changes to the package or module. 

https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#commit-conventions

**Comment by gcleroux** [comment]
> 👍
> 
> Minor (sorry that I didn't link to this before):
> 
> > When adding yourself as maintainer in the same pull request, make a separate commit with the message `maintainers: add <handle>`. Add the commit before those making changes to the package or module.
> 
> https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#commit-conventions

Don't worry about it. I appreciate the pointers, it's my own damn fault for not reading the contributing README lol

**Review by amesgen — APPROVED** [review]

File: pkgs/applications/misc/bazecor/default.nix [hunk 5]
```diff
@@ -17,12 +17,15 @@ appimageTools.wrapAppImage rec {
     # Workaround for https://github.com/Dygmalab/Bazecor/issues/370
     postExtract = ''
       substituteInPlace \
-        $out/usr/lib/bazecor/resources/app/.webpack/main/index.js \
-        --replace \
+        $out/usr/lib/${pname}/resources/app/.webpack/main/index.js \
```
Comment (Aleksanaa) [inline pkgs/applications/misc/bazecor/default.nix:20]:
```suggestion
        $out/usr/lib/bazecor/resources/app/.webpack/main/index.js \
```

Don't reference `pname` everywhere. It's of no use.

Comment (gcleroux) [reply pkgs/applications/misc/bazecor/default.nix:20]: > Don't reference `pname` everywhere. It's of no use.

This is a bit overzealous and opinionated for no valid reason other than personal preference as far as I can tell. 
I'm happy to adjust the PR to fit the contributing guidelines, but I don't think this is discouraged.

I can find many example in `nixpkgs` where exactly this approach is used:

https://github.com/NixOS/nixpkgs/blob/33d1e753c82ffc557b4a585c77de43d4c922ebb5/pkgs/os-specific/linux/uhk-agent/default.nix#L14-L19
https://github.com/NixOS/nixpkgs/blob/33d1e753c82ffc557b4a585c77de43d4c922ebb5/pkgs/by-name/co/codux/package.nix#L18-L25
https://github.com/NixOS/nixpkgs/blob/33d1e753c82ffc557b4a585c77de43d4c922ebb5/pkgs/tools/networking/requestly/default.nix#L17-L23

All of this to say, it's just a matter of where I took inspiration from and what you personally prefer and I don't feel like this debate is of any value.

Comment (Aleksanaa) [reply pkgs/applications/misc/bazecor/default.nix:20]: Okay, I feel sorry about that.

Comment (gcleroux) [reply pkgs/applications/misc/bazecor/default.nix:20]: That's ok, no hard feelings! ;)

Comment (Aleksanaa) [reply pkgs/applications/misc/bazecor/default.nix:20]: This is raised in this issue: https://github.com/NixOS/nixpkgs/issues/277994

Comment (gcleroux) [reply pkgs/applications/misc/bazecor/default.nix:20]: > This is raised in this issue: #277994

Which has been opened months ago without any new updates since January. Like I stated earlier, I think this is a waste of time and energy and this whole debate is quite pointless.

I will update the PR according to #277994 because I feel like we wasted entirely too much time and energy over something so irrelevant.

**Review by amesgen — APPROVED** [review]
Thanks for your work here @gcleroux!

File: pkgs/applications/misc/bazecor/default.nix [hunk 6]
```diff
@@ -35,14 +38,18 @@ appimageTools.wrapAppImage rec {
...
+    source "${makeWrapper}/nix-support/setup-hook"
+    wrapProgram $out/bin/bazecor \
+      --add-flags "\''${NIXOS_OZONE_WL:+\''${WAYLAND_DISPLAY:+--ozone-platform-hint=auto --enable-features=WaylandWindowDecorations}}"
 
+    install -m 444 -D ${src}/Bazecor.desktop -t $out/share/applications
     install -m 444 -D ${src}/bazecor.png -t $out/share/pixmaps
 
     mkdir -p $out/lib/udev/rules.d
-    ln -s --target-directory=$out/lib/udev/rules.d ${./10-dygma.rules}
+    ln -s ${./10-dygma.rules} $out/lib/udev/rules.d/10-dygma.rules
```
Comment (Aleksanaa) [inline pkgs/applications/misc/bazecor/default.nix:49]:
```suggestion
    install -Dm444 ${./10-dygma.rules} $out/lib/udev/rules.d/10-dygma.rules
```

Otherwise there will be one more output path in store

**Review by amesgen — APPROVED** [review]

File: pkgs/applications/misc/bazecor/default.nix [hunk 7]
```diff
@@ -35,14 +38,18 @@ appimageTools.wrapAppImage rec {
   # to allow non-root modifications to the keyboards.
 
   extraInstallCommands = ''
-    install -m 444 -D ${src}/Bazecor.desktop -t $out/share/applications
-    substituteInPlace $out/share/applications/Bazecor.desktop \
-      --replace 'Exec=Bazecor' 'Exec=bazecor'
+    source "${makeWrapper}/nix-support/setup-hook"
```
Comment (SuperSandro2000) [inline pkgs/applications/misc/bazecor/default.nix:41]: Can't we add makeWrapper to nativeBuildInputs?

Comment (gcleroux) [reply pkgs/applications/misc/bazecor/default.nix:41]: I could be wrong, but iirc I don't think there's a nativeBuildInputs with appimageTools. 

If it's not available, we would need to rework the PR completely to use `mkDerivation` instead and it would make more sense to use apimageTools imo.

**Comment by gcleroux** [comment]
What is currently preventing this from getting merged?

It's been idling around for a few weeks now.

**Review by Aleksanaa — APPROVED** [review]

File: pkgs/applications/misc/bazecor/default.nix (hunk 1 above)
[the remark this answers is not in the document]
Comment (zh4ngx) [reply pkgs/applications/misc/bazecor/default.nix]: I ran into this type of problem again packaging `beeper`, which requires `# passthru src to make nix-update work`, and it was a real head-scratcher. Because `wrapType2` calls `extract`, it makes sense to me to pass `postExtract` through.
